engine/api/migrate/key.go

```diff
@@ -0,0 +1,277 @@
+	if errE != nil {
+		return
+	}
+	for _, id := range envIds {
+		if err := migrateEnvironment(db, id); err != nil {
+			continue
+		}
+	}
+}
+func loadProjectIDs(db gorp.SqlExecutor) ([]int64, error) {
+	query := `SELECT project_id from project_variable where var_type = 'key'`
+	rows, errQ := db.Query(query)
+	if errQ != nil {
+		log.Warning("loadProjectIDs> Cannot load project ids: %s", errQ)
+		return nil, errQ
+	}
+	var ids []int64
+	defer rows.Close()
+	for rows.Next() {
+		var id int64
+		if err := rows.Scan(&id); err != nil {
+			log.Warning("loadProjectIDs> Cannot key next id: %s", err)
+			return nil, err
+		}
+		ids = append(ids, id)
+	}
+	return ids, nil
+}
+
+func migrateProject(db *gorp.DbMap, projID int64, store cache.Store, u *sdk.User) error {
+	tx, errT := db.Begin()
+	if errT != nil {
+		log.Warning("migrateProject> Cannot start transaction: %s", errT)
+		return errT
+	}
+	defer tx.Rollback()
+
+	proj, errP := project.LoadByID(tx, store, projID, u, project.LoadOptions.WithLockNoWait, project.LoadOptions.WithVariablesWithClearPassword, project.LoadOptions.WithKeys)
+	if errP != nil {
+		log.Warning("migrateProject> Cannot get project %d: %s", projID, errP)
+		return errP
+	}
+
+	// find private key
+	keys := findKeyPair(proj.Variable)
+	for _, k := range keys {
+		keyName := fmt.Sprintf("proj-%s", k.private.Name)
+		found := false
+		for _, k := range proj.Keys {
+			if k.Name == keyName {
+				found = true
+				break
+			}
+		}
+
+		if !found {
+			projectKey := sdk.ProjectKey{
+				ProjectID: projID,
+				Builtin:   false,
+				Key: sdk.Key{
+					Public:  k.public.Value,
+					Private: k.private.Value,
+					Name:    keyName,
+					Type:    sdk.KeySSHParameter,
+				},
+			}
+			if errK := project.InsertKey(tx, &projectKey); errK != nil {
+				log.Warning("migrateProject> Cannot insert project key %s: %s", k.private.Name, errK)
+				return errK
+			}
+		}
+	}
+
+	return tx.Commit()
+}
+
+func loadApplicationIDs(db gorp.SqlExecutor) ([]int64, error) {
+	query := `SELECT application_id from application_variable where var_type = 'key'`
+	rows, errQ := db.Query(query)
+	if errQ != nil {
+		log.Warning("loadApplicationIDs> Cannot load application ids: %s", errQ)
+		return nil, errQ
+	}
+	var ids []int64
+	defer rows.Close()
+	for rows.Next() {
+		var id int64
+		if err := rows.Scan(&id); err != nil {
+			log.Warning("loadApplicationIDs> Cannot key next id: %s", err)
+			return nil, err
+		}
+		ids = append(ids, id)
+	}
+	return ids, nil
+}
+
+func migrateApplication(db *gorp.DbMap, appID int64, store cache.Store, u *sdk.User) error {
+	tx, errT := db.Begin()
+	if errT != nil {
+		log.Warning("migrateApplication> Cannot start transaction: %s", errT)
+		return errT
+	}
+
+	app, errA := application.LoadAndLockByID(tx, store, appID, u, application.LoadOptions.WithVariablesWithClearPassword, application.LoadOptions.WithKeys)
+	if errA != nil {
+		log.Warning("migrateApplication> Cannot load application %d: %s", appID, errA)
+	}
```

Comment:
missing return ? because if there is an error the app after will be nil, and it will panic